HW2/P5/physics.pyx

```diff
@@ -69,22 +72,38 @@ cdef void sub_update(FLOAT[:, ::1] XY,
     ############################################################
     # SUBPROBLEM 2: use the grid values to reduce the number of other
     # objects to check for collisions.
-    for j in range(i + 1, count):
-        XY2 = &(XY[j, 0])
-        V2 = &(V[j, 0])
-        if overlapping(XY1, XY2, R):
-            # SUBPROBLEM 4: Add locking
-            if not moving_apart(XY1, V1, XY2, V2):
-                collide(XY1, V1, XY2, V2)
-
-            # give a slight impulse to help separate them
-            for dim in range(2):
-                V2[dim] += eps * (XY2[dim] - XY1[dim])
+
+    grid_x = <int> (XY[i, 0]/grid_spacing)
+    grid_y = <int> (XY[i, 1]/grid_spacing)
+
+    # check 2 squares away in either direction
+    for x_val in range(max(0, grid_x-2), min(grid_size, grid_x+3)):
+        for y_val in range(max(0, grid_y-2), min(grid_size, grid_y+3)):
+            if x_val == grid_x and y_val == grid_y:
+                continue
+            
+            j = Grid[x_val, y_val]
+            # The j < i deals with deadlock since we only need to compare one direction.
+            if j == -1 or j < i:
+                continue
+            acquire(&locks[j])
+            XY2 = &(XY[j, 0])
+            V2 = &(V[j, 0])
+            if overlapping(XY1, XY2, R):
+                if not moving_apart(XY1, V1, XY2, V2):
+                    collide(XY1, V1, XY2, V2)
+                # give a slight impulse to help separate them
+                for dim in range(2):
+                    V2[dim] += eps * (XY2[dim] - XY1[dim])
+            release(&locks[j])
+    release(&locks[i])
+
```

Comment:
Are you sure you always access the locks in the same order? (to avoid deadlocks)
My guess is that you locking makes the computation much slower because of a reason related to this.